ssl/quic/qlog_event_helpers.c

```diff
@@ -0,0 +1,633 @@
+                goto unknown;
+
+            QLOG_STR("frame_type", "data_blocked");
+            QLOG_U64("limit", x);
+        }
+        break;
+    case OSSL_QUIC_FRAME_TYPE_STREAM_DATA_BLOCKED:
+        {
+            uint64_t stream_id, x;
+
+            if (!ossl_quic_wire_decode_frame_stream_data_blocked(pkt,
+                                                                 &stream_id,
+                                                                 &x))
+                goto unknown;
+
+            QLOG_STR("frame_type", "stream_data_blocked");
+            QLOG_U64("stream_id", stream_id);
+            QLOG_U64("limit", x);
+        }
+        break;
+    case OSSL_QUIC_FRAME_TYPE_STREAMS_BLOCKED_BIDI:
+    case OSSL_QUIC_FRAME_TYPE_STREAMS_BLOCKED_UNI:
+        {
+            uint64_t x;
+
+            if (!ossl_quic_wire_decode_frame_streams_blocked(pkt, &x))
+                goto unknown;
+
+            QLOG_STR("frame_type", "streams_blocked");
+            QLOG_STR("stream_type",
+                     frame_type == OSSL_QUIC_FRAME_TYPE_STREAMS_BLOCKED_BIDI
+                        ? "bidirectional" : "unidirectional");
+            QLOG_U64("limit", x);
+        }
+        break;
+    case OSSL_QUIC_FRAME_TYPE_NEW_TOKEN:
+        {
+            const unsigned char *token;
+            size_t token_len;
+
+            if (!ossl_quic_wire_decode_frame_new_token(pkt, &token, &token_len))
+                goto unknown;
+
+            QLOG_STR("frame_type", "new_token");
+            QLOG_BEGIN("token");
+                QLOG_BEGIN("raw");
+                    QLOG_BIN("data", token, token_len);
+                QLOG_END();
+            QLOG_END();
+        }
+        break;
+    default:
+unknown:
+        QLOG_STR("frame_type", "unknown");
+        QLOG_U64("frame_type_value", frame_type);
+
+        /*
+         * Can't continue scanning for frames in this case as the frame length
+         * is unknown. We log the entire body the rest of the packet payload
+         * as the raw data of the frame.
+         */
+        QLOG_BEGIN("raw");
+            QLOG_BIN("data", PACKET_data(&orig_pkt),
+                     PACKET_remaining(&orig_pkt));
+        QLOG_END();
+        ignore_res(PACKET_forward(pkt, PACKET_remaining(pkt)));
+        break;
+    }
+
+    return 1;
+}
+
+static void log_frame(QLOG *qlog_instance, PACKET *pkt,
+                      uint64_t *need_skip)
+{
+    size_t rem_before, rem_after;
+
+    rem_before = PACKET_remaining(pkt);
+
+    if (!log_frame_actual(qlog_instance, pkt, need_skip))
+        return;
+
+    rem_after = PACKET_remaining(pkt);
+    QLOG_U64("length", rem_before - rem_after);
+}
+
+static int log_frames(QLOG *qlog_instance,
+                      const OSSL_QTX_IOVEC *iovec,
+                      size_t num_iovec)
+{
+    size_t i;
+    PACKET pkt;
+    uint64_t need_skip = 0;
+
+    for (i = 0; i < num_iovec; ++i) {
+        if (!PACKET_buf_init(&pkt, iovec[i].buf, iovec[i].buf_len))
```

Comment:
Are iovecs guaranteed to always hold complete frames...i.e. we never split a frame across iovecs?

Reply:
We do split a frame across iovecs. But a frame's *headers* are never split across iovecs. The loops on line 536 and 540 handle this correctly for everything we currently generate.